juju/juju
# Merge 2.2 into develop (storage fixes)

## Description of change

Merge storage fixes from 2.2 into develop. Things have changed a bit: the scope of a filesystem is now always the same as the backing volume, and when you destroy a unit, it leaves detachable storage in the model.

So we now also cater for detachable, volume-backed filesystems by short-circuit removing their attachments, without destroying the filesystem. We also update the status of filesystems for which we remove the attachments (i.e. set to "detached").

## QA steps

1. juju bootstrap aws
2. juju deploy jenkins --storage jenkins=ebs
(wait)
3. juju ssh 0
cd /srv/mnt/jenkins
(sit there to prevent filesystem from being unmounted)
4. juju remove-application jenkins
(wait; machine should be torn down, filesystem and volume should be left in the model; status should be "detached", with no error message)
5. juju destroy-model --destroy-storage
(filesystem and volume should be destroyed)

## Documentation changes

None.

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1722818
